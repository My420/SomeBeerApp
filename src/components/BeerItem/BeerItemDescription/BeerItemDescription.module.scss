@import '../../../sass/variables.scss';

.tabList {
  display: flex;
  width: 100%;
  margin-bottom: 20px;

  list-style: none;
}

.tabItem {
  padding: 10px 30px;
  text-transform: uppercase;
  background-color: $menu-bar-color;
  border: 2px solid transparent;
  cursor: pointer;
}

.tabItem:focus {
  outline: none;
  border: 2px solid $menu-hover-color;
}

.tabItem[data-active='true'] {
  background-color: $chip-color;
  cursor: auto;
}

.foodList {
  padding-left: 30px;
}

@media (max-width: $Tablet9-max-width) {
  .tabItem {
    font-size: 16px;
    line-height: 20px;
  }

  .foodItem,
  .description {
    font-size: 14px;
    line-height: 18px;
  }
}

@media (max-width: 640px) {
  .tabItem {
    font-size: 14px;
    line-height: 18px;
  }
}

@media (max-width: $Tablet7-max-width) {
  .tabItem {
    padding: 5px 15px;
  }
}
