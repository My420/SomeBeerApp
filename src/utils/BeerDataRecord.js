import { Record } from 'immutable';

const BeerDataRecord = Record(
  {
    id: null,
    name: null,
    tagline: null,
    first_brewed: null,
    description: null,
    image_url: null,
    abv: null,
    ibu: null,
    target_fg: null,
    target_og: null,
    ebc: null,
    srm: null,
    ph: null,
    attenuation_level: null,
    volume: {},
    boil_volume: {},
    method: {},
    ingredients: {},
    food_pairing: [],
    brewers_tips: null,
    contributed_by: null
  },
  'BeerDataRecord'
);

export default BeerDataRecord;
