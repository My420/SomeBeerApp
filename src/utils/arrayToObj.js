/* eslint no-param-reassign: "error" */

const arrayToObj = arr => {
  return arr.reduce((prevValue, element) => {
    prevValue[element.id] = element;
    return prevValue;
  }, {});
};

export default arrayToObj;
