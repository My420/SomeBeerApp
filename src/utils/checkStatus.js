const checkStatus = response => {
  const responseStatus = +response.status;
  if (responseStatus >= 200 && responseStatus < 300) {
    return response;
  }
  throw new Error(response.statusText);
};

export default checkStatus;
