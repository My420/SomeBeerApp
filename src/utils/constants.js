export const SHOP_NAME = 'Some Beer Shop';

export const LINK = {
  HOME: '/',
  CATALOG: '/catalog',
  ROULETTE: '/roulette',
  ABOUT: '/about',
  FAVORITE: '/favorite',
  CART: '/cart'
};

export const NAVIGATION_LINKS = [
  [LINK.HOME, 'Home'],
  [LINK.CATALOG, 'Catalog'],
  [LINK.ROULETTE, 'Roulette!'],
  [LINK.FAVORITE, 'Favorite'],
  [LINK.ABOUT, 'About'],
  [LINK.CART, 'Cart']
];

export const BROWSER_WIDTH_RATIO = {
  LPC: 4,
  MPC: 3,
  TABLET_9: 2,
  TABLET_7: 1,
  MOBILE: 0
};

export const BROWSER_MIN_WIDTH = {
  LPC: 1200,
  MPC: 993,
  TABLET_9: 768,
  TABLET_7: 480,
  MOBILE: 320
};

export const SOCIAL_LINK = {
  GITHUB: 'https://github.com/My420',
  EMAIL: 'mailto:klinovitsky.aleksey@gmail.com',
  SKYPE: 'skype:live:klinovitsky.aleksey?chat'
};

export const AUTHOR = 'Klinovitsky Aleksey';

export const POPULAR_BEER__GET_REQUEST = 'POPULAR_BEER__GET_REQUEST';
export const POPULAR_BEER__GET_SUCCESS = 'POPULAR_BEER__GET_SUCCESS';
export const POPULAR_BEER__GET_FAILURE = 'POPULAR_BEER__GET_FAILURE';

export const API_HOST = 'https://api.punkapi.com/v2/';

export const POPULAR_BEER_AMOUNT = 8;

export const LOAD_ERROR_MESSAGE =
  'Could not download data, please refresh the page';

export const ERROR_TITLE = 'Oops! Something went wrong.';
export const ERROR_ACTION = 'Please refresh page or try again later.';
