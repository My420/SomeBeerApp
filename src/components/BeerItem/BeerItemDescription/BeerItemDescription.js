import React from 'react';
import PropTypes from 'prop-types';
import styles from './BeerItemDescription.module.scss';
import ListGroup from '../../ListGroup/ListGroup';
import ListGroupItem from '../../ListGroupItem/ListGroupItem';
import { ENTER_CHAR_CODE } from '../../../utils/constants';

class BeerItemDescription extends React.Component {
  state = { current: 'Description' };

  onTabCLick = evt => {
    const { tab } = evt.target.dataset;
    const { current } = this.state;
    if (current !== tab) {
      this.setState({ current: tab });
    }
  };

  onKeyPress = evt => {
    const key = evt.charCode;
    if (key === ENTER_CHAR_CODE) {
      this.onTabCLick(evt);
    }
  };

  getFoodDescription = () => {
    const {
      description: { Food: food }
    } = this.props;

    const foodDescription = food.map(elem => {
      return (
        <ListGroupItem className={styles.foodItem} key={elem.slice(0, 8)}>
          {elem}
        </ListGroupItem>
      );
    });

    return <ListGroup className={styles.foodList}>{foodDescription}</ListGroup>;
  };

  getDescription = () => {
    const { description } = this.props;
    const { current } = this.state;
    if (current === 'Food') {
      return this.getFoodDescription();
    }

    return <p className={styles.description}>{description[current]}</p>;
  };

  getTabs = () => {
    const { description } = this.props;
    const { current } = this.state;
    const tabs = Object.keys(description).map(tab => {
      const isActive = tab === current;
      return (
        <ListGroupItem
          className={styles.tabItem}
          key={tab}
          data-tab={tab}
          data-active={isActive}
          onClick={this.onTabCLick}
          onKeyPress={this.onKeyPress}
          tabIndex="0"
        >
          {tab}
        </ListGroupItem>
      );
    });

    return tabs;
  };

  render() {
    return (
      <React.Fragment>
        <ListGroup className={styles.tabList}>{this.getTabs()}</ListGroup>
        {this.getDescription()}
      </React.Fragment>
    );
  }
}

BeerItemDescription.propTypes = {
  description: PropTypes.shape({
    Description: PropTypes.string.isRequired,
    Food: PropTypes.array.isRequired,
    Tips: PropTypes.string.isRequired
  }).isRequired
};

export default BeerItemDescription;
