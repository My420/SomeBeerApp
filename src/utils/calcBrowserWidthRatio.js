import { BROWSER_MIN_WIDTH, BROWSER_WIDTH_RATIO } from './constants';

const calcBrowserWidthRatio = width => {
  if (width > BROWSER_MIN_WIDTH.LPC) {
    return BROWSER_WIDTH_RATIO.LPC;
  }
  if (width > BROWSER_MIN_WIDTH.MPC) {
    return BROWSER_WIDTH_RATIO.MPC;
  }
  if (width > BROWSER_MIN_WIDTH.TABLET) {
    return BROWSER_WIDTH_RATIO.TABLET_9;
  }
  if (width > BROWSER_MIN_WIDTH.MOBILE) {
    return BROWSER_WIDTH_RATIO.TABLET_7;
  }
  return BROWSER_WIDTH_RATIO.MOBILE;
};

export default calcBrowserWidthRatio;
