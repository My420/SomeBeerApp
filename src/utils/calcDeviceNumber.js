import { DEVICE_MIN_WIDTH, DEVICE_NUMBER } from './constants';

const calcDeviceNumber = width => {
  if (width > DEVICE_MIN_WIDTH.LPC) {
    return DEVICE_NUMBER.LPC;
  }
  if (width > DEVICE_MIN_WIDTH.MPC) {
    return DEVICE_NUMBER.MPC;
  }
  if (width > DEVICE_MIN_WIDTH.TABLET) {
    return DEVICE_NUMBER.TABLET;
  }
  return DEVICE_NUMBER.MOBILE;
};

export default calcDeviceNumber;
