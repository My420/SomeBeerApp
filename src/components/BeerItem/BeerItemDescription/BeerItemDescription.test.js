import React from 'react';
import { shallow } from 'enzyme';
import * as fc from 'fast-check';
import BeerItemDescription from './BeerItemDescription';
import { ENTER_CHAR_CODE } from '../../../utils/constants';

describe('test <BeerItemDescription /> component', () => {
  const description = {
    Description: 'Description',
    Food: ['food1', 'food2'],
    Tips: 'Tips'
  };

  describe('tabs should change on tab click or Enter press', () => {
    const wrapper = shallow(<BeerItemDescription description={description} />);
    const keys = Object.keys(description);

    test("should correct change state on tab click'", () => {
      fc.assert(
        fc.property(fc.integer(0, keys.length - 1), i => {
          const key = keys[i];
          const tab = wrapper.find(`ListGroupItem[data-tab='${key}']`);
          const event = { target: { dataset: { tab: key } } };
          tab.simulate('click', event);
          expect(wrapper.state('current')).toBe(key);
        })
      );
    });

    test("should correct change state on press key Enter on tab'", () => {
      fc.assert(
        fc.property(fc.integer(0, keys.length - 1), i => {
          const key = keys[i];
          const tab = wrapper.find(`ListGroupItem[data-tab='${key}']`);
          const event = {
            target: { dataset: { tab: key } },
            charCode: ENTER_CHAR_CODE
          };
          tab.simulate('keyPress', event);
          expect(wrapper.state('current')).toBe(key);
        })
      );
    });
  });

  describe('test <BeerItemDescription /> shuld render Description tab first', () => {
    const wrapper = shallow(<BeerItemDescription description={description} />);

    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test("should have state.current === 'Description' by default", () => {
      expect(wrapper.state('current')).toBe('Description');
    });

    test('should correct render tabs', () => {
      const keys = Object.keys(description);
      const tabs = wrapper.find('ListGroupItem.tabItem');
      expect(tabs).toHaveLength(keys.length);
      for (let i = 0; i < keys.length; i += 1) {
        expect(tabs.at(i).prop('children')).toBe(keys[i]);
      }
    });

    test("should correct render tab 'Description' description", () => {
      expect(wrapper.find('p.description')).toHaveLength(1);
      expect(wrapper.find('p.description').text()).toBe(
        description.Description
      );
    });
  });

  describe("test <BeerItemDescription /> shuld render 'Tips' tab", () => {
    const wrapper = shallow(<BeerItemDescription description={description} />);
    wrapper.setState({ current: 'Tips' });

    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test("should have state.current === 'Tips'", () => {
      expect(wrapper.state('current')).toBe('Tips');
    });

    test('should correct render tabs', () => {
      const keys = Object.keys(description);
      const tabs = wrapper.find('ListGroupItem.tabItem');
      expect(tabs).toHaveLength(keys.length);
      for (let i = 0; i < keys.length; i += 1) {
        expect(tabs.at(i).prop('children')).toBe(keys[i]);
      }
    });

    test("should correct render tab 'Tips' description", () => {
      expect(wrapper.find('p.description')).toHaveLength(1);
      expect(wrapper.find('p.description').text()).toBe(description.Tips);
    });
  });

  describe("test <BeerItemDescription /> shuld render 'Food' tab", () => {
    const wrapper = shallow(<BeerItemDescription description={description} />);
    wrapper.setState({ current: 'Food' });

    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test("should have state.current === 'Food'", () => {
      expect(wrapper.state('current')).toBe('Food');
    });

    test('should correct render tabs', () => {
      const keys = Object.keys(description);
      const tabs = wrapper.find('ListGroupItem.tabItem');
      expect(tabs).toHaveLength(keys.length);
      for (let i = 0; i < keys.length; i += 1) {
        expect(tabs.at(i).prop('children')).toBe(keys[i]);
      }
    });

    test("should correct render tab 'Food' description", () => {
      const { Food } = description;
      const elems = wrapper.find('ListGroupItem.foodItem');
      expect(elems).toHaveLength(Food.length);
      for (let i = 0; i < Food.length; i += 1) {
        expect(elems.at(i).prop('children')).toBe(Food[i]);
      }
    });
  });
});
