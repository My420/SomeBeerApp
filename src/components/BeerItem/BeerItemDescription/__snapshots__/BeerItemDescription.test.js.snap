// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <BeerItemDescription /> component test <BeerItemDescription /> shuld render 'Food' tab should match snapshot 1`] = `
<Fragment>
  <ListGroup
    className="tabList"
    tag="ul"
  >
    <ListGroupItem
      className="tabItem"
      data-active={false}
      data-tab="Description"
      key="Description"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Description
    </ListGroupItem>
    <ListGroupItem
      className="tabItem"
      data-active={true}
      data-tab="Food"
      key="Food"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Food
    </ListGroupItem>
    <ListGroupItem
      className="tabItem"
      data-active={false}
      data-tab="Tips"
      key="Tips"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Tips
    </ListGroupItem>
  </ListGroup>
  <ListGroup
    className="foodList"
    tag="ul"
  >
    <ListGroupItem
      className="foodItem"
      key="food1"
    >
      food1
    </ListGroupItem>
    <ListGroupItem
      className="foodItem"
      key="food2"
    >
      food2
    </ListGroupItem>
  </ListGroup>
</Fragment>
`;

exports[`test <BeerItemDescription /> component test <BeerItemDescription /> shuld render 'Tips' tab should match snapshot 1`] = `
<Fragment>
  <ListGroup
    className="tabList"
    tag="ul"
  >
    <ListGroupItem
      className="tabItem"
      data-active={false}
      data-tab="Description"
      key="Description"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Description
    </ListGroupItem>
    <ListGroupItem
      className="tabItem"
      data-active={false}
      data-tab="Food"
      key="Food"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Food
    </ListGroupItem>
    <ListGroupItem
      className="tabItem"
      data-active={true}
      data-tab="Tips"
      key="Tips"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Tips
    </ListGroupItem>
  </ListGroup>
  <p
    className="description"
  >
    Tips
  </p>
</Fragment>
`;

exports[`test <BeerItemDescription /> component test <BeerItemDescription /> shuld render Description tab first should match snapshot 1`] = `
<Fragment>
  <ListGroup
    className="tabList"
    tag="ul"
  >
    <ListGroupItem
      className="tabItem"
      data-active={true}
      data-tab="Description"
      key="Description"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Description
    </ListGroupItem>
    <ListGroupItem
      className="tabItem"
      data-active={false}
      data-tab="Food"
      key="Food"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Food
    </ListGroupItem>
    <ListGroupItem
      className="tabItem"
      data-active={false}
      data-tab="Tips"
      key="Tips"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex="0"
    >
      Tips
    </ListGroupItem>
  </ListGroup>
  <p
    className="description"
  >
    Description
  </p>
</Fragment>
`;
